tests/integration_tests/datasources/test_nocloud.py

```diff
@@ -88,3 +95,237 @@ def test_nocloud_seedfrom_vendordata(client: IntegrationInstance):
+                #
+                # we could check out a tag, but the project hasn't
+                # been updated in 2 years
+                #
+                # instructions from https://github.com/FiloSottile/mkcert
+                assert client.execute(
+                    "git clone https://github.com/FiloSottile/mkcert && "
+                    "cd mkcert && "
+                    "go build -ldflags "
+                    '"-X main.Version=$(git describe --tags)"'
+                ).ok
+
+                # giddyup
+                assert client.execute(
+                    "ln -s $HOME/mkcert/mkcert /usr/local/bin/mkcert"
+                ).ok
+
+            # more palatable than openssl commands
+            assert client.execute(
+                "mkcert -install -cert-file /cert.pem -key-file /key.pem "
+                "localhost 127.0.0.1 0.0.0.0 ::1"
+            ).ok
+
+        client.write_to_file(
+            "/lib/systemd/system/local-ftp.service",
+            dedent(
+                """\
+                [Unit]
+                Description=TESTING USE ONLY ftp server
+                Wants=cloud-init-local.service
+                DefaultDependencies=no
+
+                # we want the network up for network operations
+                # and NoCloud operates in network timeframe
+                After=systemd-networkd-wait-online.service
+                After=networking.service
+                Before=cloud-init.service
+
+                [Service]
+                Type=exec
+                ExecStart=/server.py
+
+                [Install]
+                WantedBy=cloud-init.target
+                """
+            ),
+        )
+        assert client.execute(
+            "chmod 644 /lib/systemd/system/local-ftp.service"
+        ).ok
+        assert client.execute("systemctl enable local-ftp.service").ok
+        assert client.execute("mkdir /home/anonymous").ok
+
+        client.write_to_file(
+            "/user-data",
+            dedent(
+                """\
+                #cloud-config
+
+                hostname: ftp-bootstrapper
+                """
+            ),
+        )
+        client.write_to_file(
+            "/meta-data",
+            dedent(
+                """\
+                instance-id: ftp-instance
+                """
+            ),
+        )
+        client.write_to_file("/vendor-data", "\n")
+
+        # set the kernel commandline, reboot with it
+        override_kernel_cmdline(cmdline, client)
+
+    def test_nocloud_ftp_unencrypted_server_succeeds(
+        self, client: IntegrationInstance
+    ):
+        """check that ftp:// succeeds to unencrypted ftp server
+
+        this mode allows administrators to choose unencrypted ftp,
+        at their own risk
+        """
+        cmdline = "ds=nocloud;seedfrom=ftp://0.0.0.0:2121"
+        self._boot_with_cmdline(cmdline, client)
+        verify_clean_boot(client, ignore_warnings=self.expected_warnings)
```

Comment:
Can we also verify here (and the other tests) that the userdata worked correctly? I.e., verify the hostname or add a bootmcd to the userdata to touch a file and show the file was actually touched.

Reply:
done